### PR TITLE
fix #2325 Add method to snapshot factory+global schedulers, used by VTS

### LINKED ISSUE #2325: Explore a way to keep current `Schedulers` live while temporarily replacing the `Factory`
## Motivation
Some tests might run with the `VirtualTimeScheduler` while other parts of the test infrastructure might also rely on decorating `Schedulers` (via the `Schedulers.Factory`) and keeping these decorated schedulers live for the duration of eg. a test class.

Unfortunately, setting up the `VirtualTimeScheduler` is destructive: it shuts down previously set schedulers.

## Desired solution
Add a way to capture the current live `Schedulers` and restore them.

This probably involves an object slightly different from the `Schedulers.Factory`, which responsability is to _create_ new `Schedulers`. Here we're talking about capturing a set of existing `Scheduler` instances (or rather likely `CachedScheduler`) in a black box object that can be passed back to `Schedulers` to restore said instances. The current `Factory` should probably also be saved.

## Considered alternatives
A special flavor of `Schedulers.Factory`, but that is impractical since the cached live schedulers are held by the `Schedulers` class directly, as package private `CachedScheduler` instances in atomic references.

## Challenges
There is no way around adding methods and specific support for that directly on `Schedulers`, since it is the class that holds live instances of `Scheduler`.

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

**Comment by simonbasle** [comment]
@reactor/core-team I initially opened the issue against 3.3, but I think the change in behavior, while subtle, is enough to warrant targeting 3.4 instead. wdyt? cc @osi @rstoyanchev

**Comment by osi** [comment]
I like this approach to solving the problem!

As to whether it's worth a version bump, I leave that to you. I'd say no, as the behavior change appears isolate to usages of reactor-test.

File: reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java [hunk 1]
```diff
@@ -677,16 +678,62 @@ public static void disableMetrics() {
 	/**
 	 * Re-apply default factory to {@link Schedulers}
 	 */
-	public static void resetFactory(){
+	public static void resetFactory() {
 		setFactory(DEFAULT);
 	}
 
+	/**
+	 * Set the current {@link Schedulers} aside in a capturing {@link Snapshot} that
```
Comment (ericbottard) [inline reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java:686]: This should be the second sentence. The current javadoc doesn't mention that this allows setting the factory in the first place.

File: reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java [hunk 2]
```diff
@@ -677,16 +678,62 @@ public static void disableMetrics() {
...
+				CACHED_BOUNDED_ELASTIC.getAndSet(null),
+				CACHED_PARALLEL.getAndSet(null),
+				CACHED_SINGLE.getAndSet(null),
+				factory);
+		setFactory(newFactory);
+		return snapshot;
+	}
+
+	/**
+	 * @param snapshot
```
Comment (ericbottard) [inline reactor-core/src/main/java/reactor/core/scheduler/Schedulers.java:705]:
```suggestion
	 * Reset the Factory and Schedulers with the ones saved in a 
	 * previously {@link #setFactoryWithSnapshot(Factory) captured} snapshot.
	 * <p>Passing {@code null} re-applies the default factory.
```

File: reactor-test/src/main/java/reactor/test/scheduler/VirtualTimeScheduler.java [hunk 3]
```diff
@@ -77,8 +77,12 @@ public static VirtualTimeScheduler create(boolean defer) {
...
 	 * factories ONLY if no {@link VirtualTimeScheduler} is currently set. In case of scheduler creation,
 	 * there is no deferring of time operations (see {@link #create(boolean)}.
-	 * While the method is thread safe, its usually advised to execute such
-	 * wide-impact BEFORE all tested code runs (setup etc). The created scheduler is returned.
+	 * Note that prior to replacing the factories, a {@link Schedulers#setFactoryWithSnapshot(Schedulers.Factory) snapshot}
+	 * will be performed. Resetting the factory will restore said snapshot.
+	 * <p>
+	 * While this methods makes best effort to be thread safe, it is usually advised to
+	 * perform such wide-impact setup serially and BEFORE all test code runs
+	 * (setup/beforeAll/beforeClass...). The create Scheduler is returned.
```
Comment (ericbottard) [inline reactor-test/src/main/java/reactor/test/scheduler/VirtualTimeScheduler.java:85]:
```suggestion
	 * (setup/beforeAll/beforeClass...). The created Scheduler is returned.
```

File: reactor-test/src/main/java/reactor/test/scheduler/VirtualTimeScheduler.java [hunk 4]
```diff
@@ -90,8 +94,13 @@ public static VirtualTimeScheduler getOrSet() {
...
 	 * factories ONLY if no {@link VirtualTimeScheduler} is currently set. In case of scheduler creation,
 	 * there is opt-in deferring of time related operations (see {@link #create(boolean)}.
-	 * While the method is thread safe, its usually advised to execute such
-	 * wide-impact BEFORE all tested code runs (setup etc). The created scheduler is returned.
+	 * Note that prior to replacing the factories, a {@link Schedulers#setFactoryWithSnapshot(Schedulers.Factory) snapshot}
+	 * will be performed. Resetting the factory will restore said snapshot.
+	 * <p>
+	 * While this methods makes best effort to be thread safe, it is usually advised to
+	 * perform such wide-impact setup serially and BEFORE all test code runs
+	 * (setup/beforeAll/beforeClass...). The create Scheduler is returned.
```
Comment (ericbottard) [inline reactor-test/src/main/java/reactor/test/scheduler/VirtualTimeScheduler.java:102]:
```suggestion
	 * (setup/beforeAll/beforeClass...). The created Scheduler is returned.
```

**Comment by simonbasle** [comment]
we'll keep it in 3.4 for now. post RC1 we can eventually think about backporting it. @osi would you be able to check it fixes your issue? maybe recreating a repro project using reactor-core 3.4.0-SNAPSHOT ?

**Comment by osi** [comment]
@simonbasle I'll make note to try the snapshot of 3.4 against my project. I'm not sure how quick it would be to stand up a separate repro project, but I'll also keep that in mind too 😀

**Comment by simonbasle** [comment]
@osi I suggested the separate repo solution in case the effort to upgrade to 3.4 snapshots in your existing code is too high, which it might be (with netty 1.0 changes and processor changes in core 3.4)

**Comment by osi** [comment]
@simonbasle yeah that's totally fair. the code in question _is_ using processors 😀 (but hopefully in a way that is properly forward looking, interacting primarily via a `Sink`). there's also the benefit of having a reproducer repo as it could be the basis of a test that verifies this behavior.

**Pull request merged**